File: apps/backend/app/schemas/memory.py
import json
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict, field_validator

class MemoryItemCreate(BaseModel):
    source_type: str = Field(default="user_fact")
    source_id: Optional[str] = None
    content: str = Field(..., min_length=1)
    memory_tier: str = Field(..., description="CORE, RECALL, or ARCHIVAL")
    metadata: Optional[Dict[str, Any]] = None

class MemoryItemUpdate(BaseModel):
    content: Optional[str] = None
    memory_tier: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None

class MemoryItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    source_type: str
    source_id: Optional[str] = None
    content: str
    memory_tier: str
    created_at: datetime
    updated_at: datetime
    metadata_json: Optional[str] = None

class MemorySearchQuery(BaseModel):
    query: str = Field(..., min_length=1)
    tier: Optional[str] = None
    limit: int = Field(default=10, ge=1, le=100)

class CoreMemorySetRequest(BaseModel):
    user_preferences: Optional[str] = None
    active_goals: Optional[str] = None
    important_facts: Optional[str] = None

File: apps/desktop/src/services/api/memory.ts
import { apiRequest } from "./client";
import { MemoryItem } from "../../types";

export interface MemorySearchResult {
  id: string;
  content: string;
  memory_tier: 'CORE' | 'RECALL' | 'ARCHIVAL';
  source_type: string;
  source_id?: string;
  relevance_score: number;
  created_at: string;
  metadata?: Record<string, any>;
}

export const memoryApi = {
  create: (data: { content: string; memory_tier: string; source_type?: string }): Promise<MemoryItem> =>
    apiRequest<MemoryItem>("/memory", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  getCore: (): Promise<MemoryItem[]> =>
    apiRequest<MemoryItem[]>("/memory/core"),

  setCore: (data: { user_preferences?: string; active_goals?: string; important_facts?: string }): Promise<MemoryItem[]> =>
    apiRequest<MemoryItem[]>("/memory/core", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  search: (query: string, tier?: string, limit = 10): Promise<MemorySearchResult[]> =>
    apiRequest<MemorySearchResult[]>("/memory/search", {
      method: "POST",
      body: JSON.stringify({ query, tier, limit }),
    }),

  delete: (id: string): Promise<void> =>
    apiRequest<void>(`/memory/${id}`, {
      method: "DELETE",
    }),
};
